src/app/clusters/operational-state-server/operational-state-server.cpp

```diff
@@ -337,11 +337,26 @@ void Instance::HandlePauseState(HandlerContext & ctx, const Commands::Pause::Dec
     GenericOperationalError err(to_underlying(ErrorStateEnum::kNoError));
     uint8_t opState = GetCurrentOperationalState();
 
-    if (opState != to_underlying(OperationalStateEnum::kPaused) && opState != to_underlying(OperationalStateEnum::kRunning))
+    // Handle Operational State Pause-incompatible states.
+    if (opState == to_underlying(OperationalStateEnum::kStopped) || opState == to_underlying(OperationalStateEnum::kError))
     {
         err.Set(to_underlying(ErrorStateEnum::kCommandInvalidInState));
     }
-    else if (opState == to_underlying(OperationalStateEnum::kRunning))
+
+    // Handle Pause-incompatible states for derived clusters.
+    switch (mClusterId)
+    {
+    case RvcOperationalState::Id:
```

Comment:
This is almost certainly not going to be optimized out by compilers for instances of this thing that are not RvcOperationalState.  A compiler would have to prove that the class is never created anywhere with specific values of the cluster id, which is not something it knows in this translation unit.  The only thing that could make it work is if you're compiling with an LTO pass or something, but is that common for embedded devices?  Certainly our build system does not make it very feasible right now.

If we want this to really be "0-cost when not relevant", we should do one of two things:

1. We have a virtual method that subclasses that are cluster-id-specific need to implement to do this check.  Then the cost here is constant no matter how many opstate things we happen to have around, and this scales sanely to vendor-prefixed opstate derivations and so on.
2. We make the cluster id a template param.  But then you probably end up with some cost if you ever have more than one opstate instance around with different cluster ids, so this is likely not a good idea.

Reply:
I have changed the API such that all operational state clusters implement their own class. If required, they can implement additional logic by overriding virtual protected functions of the base operational state class.